compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ReachabilityAnalyzer.java

```diff
@@ -0,0 +1,685 @@
+        if (onFailClause != null) {
+            boolean currentStatementReturns = this.statementReturnsPanicsOrFails;
+            this.booleanConstCondition = BooleanConst.NOT_CONST_BOOLEAN;
+            resetStatementReturnsPanicsOrFails();
+            resetLastStatement();
+            analyzeReachability(onFailClause);
+            this.statementReturnsPanicsOrFails = currentStatementReturns;
+        }
+    }
+
+    @Override
+    public void visit(BLangRetry retryNode) {
+        boolean failureHandled = this.failureHandled;
+        checkStatementExecutionValidity(retryNode);
+        if (!this.failureHandled) {
+            this.failureHandled = retryNode.onFailClause != null;
+        }
+        analyzeReachability(retryNode.retryBody);
+        this.failureHandled = failureHandled;
+        resetLastStatement();
+        resetErrorThrown();
+        analyzeOnFailClause(retryNode.onFailClause);
+    }
+
+    @Override
+    public void visit(BLangRetryTransaction retryTransaction) {
+        analyzeReachability(retryTransaction.transaction);
+    }
+
+    @Override
+    public void visit(BLangOnFailClause onFailClause) {
+        resetLastStatement();
+        resetErrorThrown();
+        analyzeReachability(onFailClause.body);
+        resetErrorThrown();
+    }
+
+    @Override
+    public void visit(BLangFunction funcNode) {
+        resetFunction();
+        if (funcNode.flagSet.contains(Flag.NATIVE)) {
+            return;
+        }
+        if (funcNode.body != null) {
+            analyzeReachability(funcNode.body);
+            boolean isNeverReturn = types.isNeverTypeOrStructureTypeWithARequiredNeverMember
+                    (funcNode.symbol.type.getReturnType());
+            // If the return signature is nil-able, an implicit return will be added in Desugar.
+            // Hence this only checks for non-nil-able return signatures and uncertain return in the body.
+            if (!funcNode.symbol.type.getReturnType().isNullable() && !isNeverReturn &&
+                    !this.statementReturnsPanicsOrFails) {
+                Location closeBracePos = getEndCharPos(funcNode.pos);
+                this.dlog.error(closeBracePos, DiagnosticErrorCode.INVOKABLE_MUST_RETURN,
+                        funcNode.getKind().toString().toLowerCase());
+            } else if (isNeverReturn && !this.statementReturnsPanicsOrFails) {
+                this.dlog.error(funcNode.pos, DiagnosticErrorCode.THIS_FUNCTION_SHOULD_PANIC);
+            }
+        }
+
+        if (funcNode.returnTypeNode.getBType().tag == TypeTags.UNION) {
+            LinkedHashSet<BType> memberTypes = ((BUnionType) funcNode.returnTypeNode.getBType()).getMemberTypes();
+            if (memberTypes.contains(symTable.nilType) && !this.statementReturnsPanicsOrFails) {
+                this.dlog.warning(funcNode.returnTypeNode.pos,
+                        DiagnosticWarningCode.FUNCTION_SHOULD_EXPLICITLY_RETURN_A_NIL);
```

Comment:
Added this since it was mentioned in the unreachable google doc. 
```
Additional feature

Make this warning now, error later:

function foo() returns int? {
   // Should require explicit return () in this case
}
```
Is this not needed? :thinking:

Reply:
My bad, missed that bit.